locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicLocationNetworkInfoCustomizer.java

```diff
@@ -0,0 +1,472 @@
+        Iterable<String> candidates = getNodeAddressesWithMode(node);
+        return JcloudsUtil.getReachableAddresses(candidates, node.getLoginPort(), timeout, reachablePredicate);
+    }
+
+    protected Iterable<String> getNodeAddressesWithMode(NodeMetadata node) {
+        switch (getMode()) {
+        case ONLY_PRIVATE:
+            return node.getPrivateAddresses();
+        case ONLY_PUBLIC:
+            return node.getPublicAddresses();
+        case PREFER_PRIVATE:
+            return Iterables.concat(node.getPrivateAddresses(), node.getPublicAddresses());
+        case PREFER_PUBLIC:
+        default:
+            return Iterables.concat(node.getPublicAddresses(), node.getPrivateAddresses());
+        }
+    }
+
+    protected boolean isAddressResolvable(String addr) {
+        try {
+            Networking.getInetAddressWithFixedName(addr);
+            return true; // fine, it resolves
+        } catch (RuntimeException e) {
+            Exceptions.propagateIfFatal(e);
+            return false;
+        }
+    }
+
+    // --------------------------------------------------------------------------------------
+
+    protected boolean shouldCheckCredentials() {
+        return Boolean.TRUE.equals(config().get(CHECK_CREDENTIALS));
+    }
+
+    protected boolean checkCredential(
+            JcloudsLocation location, HostAndPort hostAndPort, LoginCredentials credentials,
+            ConfigBag config, boolean isWindows) {
+        try {
+            if (isWindows) {
+                location.waitForWinRmAvailable(credentials, hostAndPort, config);
+            } else {
+                location.waitForSshable(hostAndPort, ImmutableList.of(credentials), config);
+            }
+            return true;
+        } catch (IllegalStateException e) {
+            return false;
+        }
+    }
+
+    protected Iterable<LoginCredentials> getCredentialCandidates(
+            JcloudsLocation location, NodeMetadata node, ManagementAddressResolveOptions options, ConfigBag setup) {
+        LoginCredentials userCredentials = null;
+        // Figure out which login credentials to use. We only make a connection with
+        // initialCredentials when jclouds didn't do any sshing and wait for connectable is true.
+        // 0. if jclouds didn't do anything and we should wait for the machine then initial credentials is
+        //    whatever waitForSshable determines and then create the user ourselves.
+        if (options.skipJcloudsSshing() && options.expectReachable()) {
+            if (options.isWindows()) {
+                return ImmutableList.of(options.getInitialCredentials());
+            } else {
+                return location.generateCredentials(node.getCredentials(), setup.get(JcloudsLocationConfig.LOGIN_USER));
+            }
+        }
+
+        // 1. Were they configured by the user?
+        LoginCredentials customCredentials = setup.get(JcloudsLocation.CUSTOM_CREDENTIALS);
```

Comment:
`CUSTOM_CREDENTIALS` is actually in `JcloudsLocationConfig` and `JcloudsLocation` extends it - better to reference statics with the class that declares them. Here and elsewhere in this class.